test/nn/test_model_summary.py

```diff
@@ -0,0 +1,112 @@
+import torch
+from torch import Tensor, nn
+from torch_sparse import SparseTensor
+
+from torch_geometric.nn import Linear, SAGEConv, summary, to_hetero
+from torch_geometric.nn.models import GCN
+from torch_geometric.testing import withPackage
+
+
+class SAGE(torch.nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.lin1 = Linear(16, 16)
+        self.conv1 = SAGEConv(16, 32)
+        self.lin2 = Linear(32, 32)
+
+    def forward(self, x: Tensor, edge_index: Tensor) -> Tensor:
+        x = self.lin1(x).relu_()
+        x = self.conv1(x, edge_index).relu_()
+        x = self.lin2(x)
+        return x
+
+
+class ModuleDictModel(nn.Module):
+    def __init__(self):
+        super().__init__()
+        self.activations = nn.ModuleDict({
+            "lrelu": nn.LeakyReLU(),
+            "prelu": nn.PReLU()
+        })
+
+    def forward(self, x: torch.Tensor, activation_type: str) -> torch.Tensor:
+        x = self.activations[activation_type](x)
+        return x
+
+
+@withPackage('tabulate')
+def test_basic_summary() -> None:
+
+    model = GCN(32, 16, num_layers=2, out_channels=32)
+    x = torch.randn(100, 32)
+    edge_index = torch.randint(100, size=(2, 20))
+    adj_t = SparseTensor.from_edge_index(edge_index,
+                                         sparse_sizes=(100, 100)).t()
+    print(summary(model, x, edge_index))
```

Comment:
Can we actually test the output here?

Reply:
Currently I have no idea to test the output string as it's too long and complicated :(

Reply:
For the tests, I know the output is too long, so I was thinking of having some method that outputs the list (content) from `make_table` method before tabulate, so that you can check the values. However, I still believe the best way is to test the table as the library tabulate do in [test_output.py](https://github.com/zackdever/python-tabulate/blob/master/test/test_output.py).

I also suggest splitting test_summary_basic using fixtures:

```  
@pytest.fixture
def gcn(): 
    model = GCN(32, 16, num_layers=2, out_channels=32)
    x = torch.randn(100, 32)
    edge_index = torch.randint(100, size=(2, 20))
    adj_t = SparseTensor.from_edge_index(edge_index,
                                         sparse_sizes=(100, 100)).t()
    return { 'model': model,
             'x': x,
             'edge_index': edge_index,}
             'adj_t': adj_t }

    
def test_summary_edge_index(gcn) -> None:
    expected = '\n'.join(['+---------------------+--------------------+----------------+----------+',
                          '| Layer               | Input Shape        | Output Shape   | #Param   |',
                          '|---------------------+--------------------+----------------+----------|',
                          '| GCN                 | [100, 32], [2, 20] | [100, 32]      | 1,072    |',
                          '| ├─(act)ReLU         | [100, 16]          | [100, 16]      | --       |',
                          '| ├─(convs)ModuleList | --                 | --             | 1,072    |',
                          '| │    └─(0)GCNConv   | [100, 32], [2, 20] | [100, 16]      | 528      |',
                          '| │    └─(1)GCNConv   | [100, 16], [2, 20] | [100, 32]      | 544      |',
                          '+---------------------+--------------------+----------------+----------+'])

    assert summary(gcn['model'], gcn['x'], gcn['edge_index']) == expected

def test_summary_adj_t(gcn) -> None:
    expected = '\n'.join(['+---------------------+-----------------------+----------------+----------+',
                         '| Layer               | Input Shape           | Output Shape   | #Param   |',
                         '|---------------------+-----------------------+----------------+----------|',
                         '| GCN                 | [100, 32], [100, 100] | [100, 32]      | 1,072    |',
                         '| ├─(act)ReLU         | [100, 16]             | [100, 16]      | --       |',
                         '| ├─(convs)ModuleList | --                    | --             | 1,072    |',
                         '| │    └─(0)GCNConv   | [100, 32], [100, 100] | [100, 16]      | 528      |',
                         '| │    └─(1)GCNConv   | [100, 16], [100, 100] | [100, 32]      | 544      |',
                         '+---------------------+-----------------------+----------------+----------+'])

    assert summary(gcn['model'], gcn['x'], gcn['adj_t']) == expected

```

I hope that helps.

Reply:
Super! I'll update tests taking your suggestion into account, thanks!